Refactored some of OpenGLSystem

wrongu opened the pull request:
There are two parts to this PR:
- RenderTarget is treated as a class and gets its own initialization functions, instead of OpenGLSystem doing all the work
  - This includes the latest `DEPTH24_STENCIL8` option, plus the `hasDepth` parameter
- OpenGLSystem::Update is refactored into smaller functions

Note that this was mostly just moving code around, which is why there are so many additions and deletions.

Commit:
refactored RenderTarget

now a class, not a struct
rendertarget logic is now inside rendertarget methods, and called from openglsystem

Commit:
simplified Uniforms and Attributes

GLSLShader uses operator() for uniforms and operator[] for attributes. this commit removes
AddUniform and AddAttribute since it was fairly redundant

also GBuffer code is refactored

Commit:
formatting g-buffer code (comments, indents)

also added a warning if making a mesh with lighting and a non-lighting shader

Commit:
finished refactoring OpenGLSystem::Update

GLScreenQuad now has a Render() funtion with no matrices

Commit:
Merge branch 'develop' into gl_cleanup

Commit:
Using DEPTH24_STENCIL8 for RenderTarget Depth Buffer

Commit:
Merge branch 'develop' into gl_cleanup

Conflicts:
	include/systems/OpenGLSystem.h
	src/systems/OpenGLSystem.cpp

hasDepth added to RenderTargets

catageek commented on src/systems/OpenGLSystem.cpp:
```diff
@@ -576,259 +686,274 @@
+
+	void OpenGLSystem::_RenderUnlit(glm::mat4 &viewMatrix, glm::vec3 &viewPosition){
+		// Loop through and draw each GL Component.
+		for (auto eitr = this->_Components.begin(); eitr != this->_Components.end(); ++eitr) {
```
A note for future discussions:

At each iteration, we load new instructions and new data from non-contiguous memory blocks. This make the CPU cache quite inefficient if me multiply the iterations.

This is the reason why the structure of the engine should be redesigned in a real ECS, and the loop splitted in small parts for each component.

This is not specific for this PR since it is like that since the beginning.

wrongu replied:
I'm not great with cache efficiency, so can you clarify what you mean by "real ECS" and how that would make memory more contiguous?

catageek replied:
Basically, a component can store data in continuous memory blocks.

So a component can store the shaders in a vector and at each iteration, at line 864 the shader is mainly retrieved from the CPU cache since each memory access store 64 bytes of contiguous data in the cache. It may be improved with prefetch techniques.

At line 868 the Render code is different for each iteration since we change of component, so the instructions will be probably reloaded from memory. It is preferable to execute the same piece of code repeatedly. It is better to have a loop for each component with an inline function. The problem with OOP is that some components are created by inheritance and are hidden to the systems, so it is impossible to loop on components and include the implicitly created ones.

Commit:
Merge branch 'develop' into gl_cleanup

Conflicts:
	include/systems/OpenGLSystem.h
	src/systems/OpenGLSystem.cpp

Commit:
deferred rendering setup moved

no longer main's responsibility; done in OpenGLSystem::Start

Commit:
named the G-Buffer render target

renderTargets[0] is now renderTargets[this->G_BUFFER], which should be more clear

wrongu commented:
closing the issue because discussion on #128 may have more to say about this before it's merged
